Refactor `flake.nix`

aleasims opened the pull request:
Some refactoring of flake file:

- Move derivation into separate file, simplify flake file;
- Removed explicit dev shells from flake file
- Remove unnecessary step in CI workflow;
- Remove `flake-utils` from Crypto3 input (because it's not used in the current revision of Crypto3. **However** it will have to be added later).

Most of the things work the same as previously. To activate dev debug shell, run:

```
nix develop .#debug
```

Some things which may be improved in the future:
- Refine Boost library propagation properly
- Refine the list of required Go tools (they were just copy-pasted, probably we don't need all of them)

aleasims committed:
Refactor `flake.nix`

Move derivation into separate file, simplify flake file.
Remove unnecessary step in CI workflow.

aleasims committed:
Add tests of executables to CMake

Remove unneccesary CI step, which was causing rebuild of the whole derivation.

aleasims commented:
Also created targets in CMake for testing executables (block generator and assigner) runs on example data. To run these tests:

```
ninja executables_tests
```

This will compile required binaries (`assigner`, `block_gen`, ...) and run them with specified inputs.

This will allow us to run all tests inside single test derivation via Nix and avoid re-compilation of the whole derivation (CI time will decrease).

aleasims commented:
@knazarov could you please take a look at Nix part of this PR

AndreyMlashkin commented on flake.nix:
```diff
@@ -74,110 +73,28 @@
       evm-assigner = { enableDebug ? false }:
         if enableDebug then evm-assigner-pkgs.debug else evm-assigner-pkgs.default;
-
       crypto3 = nil-crypto3.packages.${system}.default;
```
use crypto3 = nil-crypto3.packages.${system}.crypto3; here

AndreyMlashkin replied:
it runs crypto3 tests on the first run now

x-mass committed:
chore: use ssh instead of https for nil inside flake.nix

ssh scheme will use ssh key instead of username/password for fetching private repo

x-mass reviewed (approved):
LGTM, except we there is no way to `nix develop` into debug with tests. Are we sure we don't want to make default check use Debug build? In this case it could be used as a target for `nix develop`. Or just return back devShells with debug+tests configuration

AndreyMlashkin committed:
update lockfile to match flake

AndreyMlashkin committed:
update nil repo

akokoshn commented:
moved to https://github.com/NilFoundation/zkEVM-framework/pull/84
